ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs

```diff
@@ -624,6 +624,20 @@ withConnectionManager ConnectionManagerArguments {
                    )
                    state
 
+        waitForOutboundDemotion
+          :: peerAddr
+          -> STM m ()
+        waitForOutboundDemotion addr = do
+          state <- readState
+          case Map.lookup addr state of
+            Nothing                  -> return ()
+            Just UnknownConnectionSt -> return ()
+            Just InboundIdleSt {}    -> return ()
+            Just InboundSt {}        -> return ()
+            Just WaitRemoteIdleSt    -> return ()
+            Just TerminatedSt        -> return ()
+            Just _                   -> retry
```

Comment:
I like my version much more, because it explicitly matches on all constructors. If we ever add a new one, we will get a warning message here.